Feature barcode splitter

gador opened the pull request:
As discussed in #285 here is the PR of a feature to split documents based on barcode separator pages. 

I'd greatly appreciate any feedback and code review.

I'll update the PR accordingly.

@paperless-ngx/backend (Python / django, database, etc.)
@paperless-ngx/test (General testing for larger PRs)

## Type of change

<!--
What type of change does your PR introduce to Paperless-ngx?
NOTE: Please check only one box!
-->

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Other (please explain)

## Checklist:

- [x] I have read & agree with the [contributing guidelines](https://github.com/paperless-ngx/paperless-ngx/blob/main/CONTRIBUTING.md).
- [x] If applicable, I have tested my code for new features & regressions on both mobile & desktop devices, using the latest version of major browsers.
- [x] If applicable, I have checked that all tests pass, see [documentation](https://paperless-ngx.readthedocs.io/en/latest/extending.html#back-end-development).
- [x] I have run all `pre-commit` hooks, see [documentation](https://paperless-ngx.readthedocs.io/en/latest/extending.html#code-formatting-with-pre-commit-hooks).
- [x] I have made corresponding changes to the documentation as needed.
- [x] I have checked my modifications for any breaking changes.


Copied from discussion:

I decided to implement solution three. I added the code to tasks.py, since it is the central point for consuming. Feature now include:

- Add an option to enable or disable the barcode splitter. 
- Make the barcode string configurable
- If enabled, documents added through the UI, consume directory or email will be split upon the barcode pages
- the new files will pe prefixed with a count and copied to the consumption dir
- the original file will be deleted, once completely split
- the new files get consumed by the default consumer, since the barcode pages have been removed

Things I did:
- Added the logic
- Added the new package dependencies to the github ci and Dockerimage
- Added unittests for the new functions
- Added the config options
- Added a paragraph to the settings documentation
- Tested with sample pdfs from the web-ui using docker-compose on the local paperless-ngx instance

Things that still need to be done:
- Add to changelog
- Add to documentation / feature list
- **Way** more extensive testing. Including email and consumption directory
- Due to the library used, this will only work with pdfs right now. A TIFF Implementation is planed, though.

gador committed:
add barcode packages

Signed-off-by: florian on nixos (Florian Brandes) <florian.brandes@posteo.de>

gador committed:
add first tests for barcode reader

Signed-off-by: florian on nixos (Florian Brandes) <florian.brandes@posteo.de>

gador committed:
working split pages

Signed-off-by: florian on nixos (Florian Brandes) <florian.brandes@posteo.de>

gador committed:
more work on barcode

Signed-off-by: florian on nixos (Florian Brandes) <florian.brandes@posteo.de>

gador committed:
skip test if dir exists

Signed-off-by: florian on nixos (Florian Brandes) <florian.brandes@posteo.de>

gador committed:
reorder imports

Signed-off-by: florian on nixos (Florian Brandes) <florian.brandes@posteo.de>

gador committed:
add trailing commas

Signed-off-by: florian on nixos (Florian Brandes) <florian.brandes@posteo.de>

gador committed:
run black

Signed-off-by: florian on nixos (Florian Brandes) <florian.brandes@posteo.de>

gador committed:
flake8

Signed-off-by: florian on nixos (Florian Brandes) <florian.brandes@posteo.de>

gador committed:
update ci.yml to include new dependency

Signed-off-by: florian on nixos (Florian Brandes) <florian.brandes@posteo.de>

gador committed:
renaming of documents to align with front end

Signed-off-by: florian on nixos (Florian Brandes) <florian.brandes@posteo.de>

gador committed:
pre-commit hooks

Signed-off-by: florian on nixos (Florian Brandes) <florian.brandes@posteo.de>

gador committed:
add config options and documentation

Signed-off-by: florian on nixos (Florian Brandes) <florian.brandes@posteo.de>

shamoon commented:
Thanks for the hard work. This does seem useful but I will note one thing which is that this adds two new system dependencies even for folks who will never use the feature. Obviously not inherently a 'bad' thing, but noting. Also its crazy to me that with all the packages we have already none of them can convert pdf to an image (I dont know the backend codebase enough to say, Im not questioning this, I just find that surprising!).

Anyway, will leave the review to the appropriate team. Thanks again for contributing

gador commented:
@shamoon 
thanks for having alook :+1: 
Actually, I first tried using the `parsers.py` `run_convert`:

https://github.com/paperless-ngx/paperless-ngx/blob/acab197a45d03c49fbac07ad756b02451f6b53ef/src/documents/parsers.py#L109-L122

The problem I encountered is that the `pyzbar` library (which decodes the barcode) requires either one of those:
` PIL / Pillow images, OpenCV / imageio / numpy ndarrays, and raw bytes` 

Imagemagick (which `run_convert` wraps), doesn't support those AFAIK. That is why I added `pdf2image`.

There is no size difference displayed by running `docker image ls`:
```
REPOSITORY              TAG                 IMAGE ID       CREATED          SIZE
paperless-ngx-dev       latest              6beab3d4b614   19 seconds ago   1.28GB
paperless_webserver     latest              d1e14147dccc   11 hours ago     1.28GB 
```

I'm personally fine with the trade off (new requirement vs. new feature), but maybe someone else has a different idea on the implementation?

shamoon commented:
Ok thanks for explaining that! Like I said, I leave the backend to you guys, just mentioning =)

stumpylog commented:
Barcode processing is something requested in ocrmypdf/OCRmyPDF#747, where the developer offered a suggestion to use a plugin for  testing for barcodes.  Maybe that could replace the `pdf2image` in some way?

gador commented:
Interesting idea.
As far as I understand it, this just describes how to hook a script in ocrmypdf. 

The script itself uses ghostscript to convert the pdf to an image. The image gets the Barcode detected by zbar.

I personally would rather use pdf2image as a python library then use a hook to an external Programm (ghostscript) to convert a pdf to an image.

gador committed:
use append method for lists

Signed-off-by: Florian Brandes <florian.brandes@posteo.de>

gador committed:
make list checks consistent

Signed-off-by: Florian Brandes <florian.brandes@posteo.de>

gador committed:
decode bytecode in string

Signed-off-by: Florian Brandes <florian.brandes@posteo.de>

gador committed:
add more type hinting

Signed-off-by: Florian Brandes <florian.brandes@posteo.de>

gador committed:
put code under CONSUMER_ENABLE_BARCODES

Signed-off-by: Florian Brandes <florian.brandes@posteo.de>

gador commented on src/documents/tasks.py:
```diff
@@ -62,6 +70,109 @@ def train_classifier():
+    else:
+        # Traverse through all the detected barcodes in image
+        for barcode in detected_barcodes:
+            if barcode.data != "":
```
I just filtered for the existence of the barcode. Maybe in the future we can add more filtering, but right now, I don't think its necessary.

stumpylog replied:
Since these are inserted by a scanner (right?) the quality of detection should already be quite high.  so I don't think we need to check for the quality either, for now

gador commented:
Hi @stumpylog 
thanks a lot for your code review! I incorporated your suggestions. Maybe you can have another look?
I still want to add a test for the `tasks.consume_file` routine. Also there is a race condition in the consumption process due to moving / renaming the file. It doesn't hinder anything, but produces a log entry. I'll have another look sometime later.

stumpylog commented:
Looks like good fixes, thank you.

- if you rebase your branch from `dev`, that should fix the failing check
- the race condition sounds like what I'm trying to fix in #483, where a file isn't actually finished before consuming begins.  if that's the case, no need to look into it
- More tests are always good.  in fact, could you add a test or 2 of files which contain another type of barcode?  And/or QR code.  Something else that the library should be able to detect, but then should be ignored.

stumpylog commented on src/paperless/settings.py:
```diff
@@ -460,6 +460,13 @@ def default_threads_per_worker(task_workers):
 
+CONSUMER_ENABLE_BARCODES = __get_boolean(
+    "PAPERLESS_CONSUMER_ENABLE_BARCODES",
+    default="YES",
```
Since this is new functionality and I'm not sure how much use it will get, I'd prefer to see this default to No/False.  We wouldn't want to start splitting if the user doesn't want us to.

shamoon replied:
Yea I debated asking for this too. If the feature works properly it shouldn't disturb documents without barcodes—right? But the other issue is if it slows down consumption even when no barcode is found. I assume it would, in which case I agree default should be `No`.

smseidl replied:
I assumed it would not be on by default when I created the discussion point.  I like where this is going, and am hoping to start testing soon.

gador replied:
I set the default to no, for now. We can still change it later, if people adopt to it. It shouldn't delay anything for people without `PATCHT` barcodes, though.

muued replied:
Well, you rasterize all pages looking for those barcodes. That will slow down processing even if you find nothing.

gador committed:
set barcode reader default to disabled

Signed-off-by: Florian Brandes <florian.brandes@posteo.de>

gador committed:
code formatting

Signed-off-by: Florian Brandes <florian.brandes@posteo.de>

stumpylog commented on src/documents/tasks.py:
```diff
@@ -62,6 +70,112 @@ def train_classifier():
         logger.warning("Classifier error: " + str(e))
 
 
+def barcode_reader(image) -> list[str]:
```
To support 3.8, I think you'll need to:
```
from typing import List
-> List[int]
```

Don't forget the `pre-commit` hooks, it will sort the imports

stumpylog commented on src/documents/tests/test_tasks.py:
```diff
@@ -89,6 +91,112 @@ def test_train_classifier(self):
+        separator_barcode = str(settings.CONSUMER_BARCODE_STRING)
+        self.assertEqual(tasks.barcode_reader(img), [separator_barcode])
+
+    def test_barcode_reader2(self):
```
I can get what the test is doing if I look at the input file, but it would be easier if the naming was descriptive for the added test cases.

test_barcode_reader_no_barcode_expected

stumpylog commented on src/documents/tests/test_tasks.py:
```diff
@@ -89,6 +91,112 @@ def test_train_classifier(self):
+        )
+        nonexistingdir = "/nowhere"
+        if os.path.isdir(nonexistingdir):
+            self.skipTest("non-existing dir exists")
```
It might be better to fail is this exists?

stumpylog commented on src/documents/tests/test_tasks.py:
```diff
@@ -89,6 +91,112 @@ def test_train_classifier(self):
+        target_file2 = os.path.join(tempdir, "patch-code-t-middle_document_1.pdf")
+        self.assertTrue(os.path.isfile(target_file1))
+        self.assertTrue(os.path.isfile(target_file2))
+
```
So all the added sample files include a PATCH-T barcode?  Could you
- generate a sample file with a QR code or other barcode to confirm the filtering by barcode type works?

gador committed:
addes tests:
- barcode-39
- barcode-128
- qr barcodes
- test for consumption

Signed-off-by: Florian Brandes <florian.brandes@posteo.de>

gador committed:
use imported List
fail test if non-existing dir exists

Signed-off-by: Florian Brandes <florian.brandes@posteo.de>

stumpylog reviewed (approved):
Now that is what I call testing!  With all these samples and the added testing, I'm good with this change.

gador commented:
rebased for current changes to CI to take effect

gador committed:
add more tests

Signed-off-by: Florian Brandes <florian.brandes@posteo.de>

gador commented:
I successfully tested:
-  a multipage scanned document
-  with other barcodes
- with PATCHT seperator pages (right and upside down)

and it worked flawlessly.

I also added two more test cases. I'll be testing it a bit more, but right now I'm quite happy. 

And due to the changes in the build process, I also rebased it to current `dev`

gador committed:
update documentation

Signed-off-by: Florian Brandes <florian.brandes@posteo.de>

smseidl commented:
I have been working to get my testing environment setup... my RPI4 was taking forever to build on.  I hope to build and test this PR this weekend.

gador commented:
I'll see whether I can provide some pre-build docker images sometime today

smseidl commented:
@gador - I'm assuming I'm doing something wrong, but my docker-compose build is failing on step 5 `Step 5/23 : RUN ./node_modules/.bin/ng build --configuration production`.  Do you have any advice?

gador commented:
Is it failing on amd64 or arm64?

I'm still trying to build (well cross build) the arm64 build. The amd64 builds for me..

smseidl commented:
> Is it failing on amd64 or arm64?
> 
> I'm still trying to build (well cross build) the arm64 build. The amd64 builds for me..

Its failing on ARM64.  Any advice on debugging this?

gador commented:
Good question. It could have something to do with #238 
Do you get any kind of error message?

smseidl commented:
> Good question. It could have something to do with #238 Do you get any kind of error message?

When I go into the container after step 4 and run the command I get this error 
```
⠙ Generating browser application bundles (phase: setup)...- ngx-file-drop [es2015/esm2015] (georgipeltekov/ngx-file-drop)
                                                            Killed
# fatal error: too many writes on closed pipe
```
The command `./node_modules/.bin/ng build --configuration production` did run if I removed the `--configuration production`.  A full build with that change to the Dockerfile still failed, but failed later.  Looks like step 14 failed with `Failed to build pikepdf` now.

step 14 is 
```
RUN apt-get update   && apt-get -y --no-install-recommends install     build-essential     && python3 -m pip install --upgrade --no-cache-dir pip wheel   && python3 -m pip install --default-timeout=1000 --upgrade --no-cache-dir supervisor   && python3 -m pip install --default-timeout=1000 --no-cache-dir -r ../requirements.txt   && apt-get -y purge build-essential   && apt-get -y autoremove --purge   && rm -rf /var/lib/apt/lists/*
```

I'm also currently online in the matrix channel if you want to chat there.

smseidl commented:
> > Good question. It could have something to do with #238 Do you get any kind of error message?
> 
> When I go into the container after step 4 and run the command I get this error
> 
> ```
> ⠙ Generating browser application bundles (phase: setup)...- ngx-file-drop [es2015/esm2015] (georgipeltekov/ngx-file-drop)
>                                                             Killed
> # fatal error: too many writes on closed pipe
> ```
> 
> The command `./node_modules/.bin/ng build --configuration production` did run if I removed the `--configuration production`. A full build with that change to the Dockerfile still failed, but failed later. Looks like step 14 failed with `Failed to build pikepdf` now.
> 
> step 14 is
> 
> ```
> RUN apt-get update   && apt-get -y --no-install-recommends install     build-essential     && python3 -m pip install --upgrade --no-cache-dir pip wheel   && python3 -m pip install --default-timeout=1000 --upgrade --no-cache-dir supervisor   && python3 -m pip install --default-timeout=1000 --no-cache-dir -r ../requirements.txt   && apt-get -y purge build-essential   && apt-get -y autoremove --purge   && rm -rf /var/lib/apt/lists/*
> ```
> 
> I'm also currently online in the matrix channel if you want to chat there.

looks like specifically this command is failing
```
python3 -m pip install --default-timeout=1000 --no-cache-dir -r ../requirements.txt
```

stumpylog commented:
I think the problem with building for aarch64 lies with the regression in the pikepdf version.  It went backwards from 5.1.0 to 5.01.  Our base image (from #238 ) provides only 5.1.0.

Most likely, if this is rebased of of `dev` and that is fixed, it will build again for all arches.

smseidl commented:
I tried a rebase this feature branch from the DEV and the issue still isn't go away.  Before the rebase I tried removing the `--no-optional` from the `npm ci` got further.  Then I got to the manage.py, script, but it failed on diago(?) missing.

stumpylog commented:
You're trying to build the Docker image, right? Not a bare metal install or build (of which I know nothing about the frontend side)?  Did you manually fix the pikepdf version in requirements.txt?

stumpylog commented on requirements.txt:
```diff
@@ -61,7 +61,8 @@ ocrmypdf==13.4.1
 pdfminer.six==20211012
-pikepdf==5.1.0
+pdf2image==1.16.0
+pikepdf==5.0.1
```
This version is now a regression from the dev branch version.

smseidl commented:
> You're trying to build the Docker image, right? Not a bare metal install or build (of which I know nothing about the frontend side)? Did you manually fix the pikepdf version in requirements.txt?

Correct - on an ARM64 machine.  I'm trying to play with the versions of pikepdf now.  What is the correct version that it should be.

stumpylog commented:
The base image provides pikepdf 5.1.0 installed.  Part of the Docker build is installing other requirements using the requirements.txt file.

If it keeps failing, could you link me a Gist of the build log?

smseidl commented:
> The base image provides pikepdf 5.1.0 installed. Part of the Docker build is installing other requirements using the requirements.txt file.
> 
> If it keeps failing, could you link me a Gist of the build log?

Thats a new one for me... Can you walk me thru it?

smseidl commented:
> The base image provides pikepdf 5.1.0 installed. Part of the Docker build is installing other requirements using the requirements.txt file.
> 
> If it keeps failing, could you link me a Gist of the build log?

Working on getting details added.  The start is here:
https://gist.github.com/smseidl/b1af20d92962876430fe885e9a40c103

smseidl commented:
> > The base image provides pikepdf 5.1.0 installed. Part of the Docker build is installing other requirements using the requirements.txt file.
> > If it keeps failing, could you link me a Gist of the build log?
> 
> Working on getting details added. The start is here: https://gist.github.com/smseidl/b1af20d92962876430fe885e9a40c103

I got 3 files in that gist now.  
* One is the straight docker-compose output after the rebase with dev of the branch
* The next, I started the last build container and re-ran the failed command within it and copied the output.
* The last (for tonight) is removing the `--no-optional` from the `npm ci` command

I hope these help...  If there could be anything missing from my system, please let me know.

stumpylog commented:
The build is trying to use pikepdf==5.0.1, built from source, which fails for a few reasons (no build tools + and upstream issue).  If you manually edit the version (change to pikepdf==5.1.0) in requirements.txt before trying to build, that will at least progress further.

shamoon commented:
Obviously stumpy is the person who will be most helpful here but a small thing is I don’t see why `--no-optional` should be removed, [the only optional npm dep at this time](https://github.com/paperless-ngx/paperless-ngx/blob/501d225f93a40472fd21455693975bd736b397e9/src-ui/package.json#L54) is a tool we use to run the e2e tests called cypress so for a docker build you shouldn’t need it. And it doesn’t install on (some) arm systems which is exactly why we made it optional. Again, just adding that small note.

smseidl commented:
> The build is trying to use pikepdf==5.0.1, built from source, which fails for a few reasons (no build tools + and upstream issue). If you manually edit the version (change to pikepdf==5.1.0) in requirements.txt before trying to build, that will at least progress further.

Ah - I read that all backwards.  I thought there was something wrong with 5.1.0 and we needed to go back.  Trying this now @stumpylog 

@shamoon - I don't know what I'm doing so you are probably right with the `no-optional` items.  I'm trying options that might change what dependencies get pulled and see what changes come in the output.  It makes sense to be as strict as we can ... I'm just trying to open it up while troubleshooting.

smseidl commented:
> > The build is trying to use pikepdf==5.0.1, built from source, which fails for a few reasons (no build tools + and upstream issue). If you manually edit the version (change to pikepdf==5.1.0) in requirements.txt before trying to build, that will at least progress further.
> 
> Ah - I read that all backwards. I thought there was something wrong with 5.1.0 and we needed to go back. Trying this now @stumpylog
> 
> @shamoon - I don't know what I'm doing so you are probably right with the `no-optional` items. I'm trying options that might change what dependencies get pulled and see what changes come in the output. It makes sense to be as strict as we can ... I'm just trying to open it up while troubleshooting.

@stumpylog - fixing that version # to 5.1.0 seems to have did the trick.  I have it up and running now.  I will start with fresh pull and containers tomorrow to ensure I don't have anything else modified in the Dockerfile.

Thanks to all for dealing with me :-)

smseidl commented:
> > > The build is trying to use pikepdf==5.0.1, built from source, which fails for a few reasons (no build tools + and upstream issue). If you manually edit the version (change to pikepdf==5.1.0) in requirements.txt before trying to build, that will at least progress further.
> > 
> > 
> > Ah - I read that all backwards. I thought there was something wrong with 5.1.0 and we needed to go back. Trying this now @stumpylog
> > @shamoon - I don't know what I'm doing so you are probably right with the `no-optional` items. I'm trying options that might change what dependencies get pulled and see what changes come in the output. It makes sense to be as strict as we can ... I'm just trying to open it up while troubleshooting.
> 
> @stumpylog - fixing that version # to 5.1.0 seems to have did the trick. I have it up and running now. I will start with fresh pull and containers tomorrow to ensure I don't have anything else modified in the Dockerfile.
> 
> Thanks to all for dealing with me :-)

I got around to testing this some more today.  The only way I am able to get this to work is by setting the pike version to 5.1.0 **and removing the `--no-optional`**.  I know this isn't ideal, but it is the only way I can get it to build.

shamoon commented:
> I got around to testing this some more today. The only way I am able to get this to work is by setting the pike version to 5.1.0 **and removing the `--no-optional`**. I know this isn't ideal, but it is the only way I can get it to build.

It sounds like the pikepdf change is correct, thats what it should be. Regarding `--no-optional`, does this branch need to be re-based to `dev`? Currently theres no issue with that on `dev`, I just re-confirmed.

That said, if the actual feature here is working then perhaps this shouldn't hold up merging if youre happy with it stumpylog, as long as there are no issues on merge, but of course this PR shouldn't affect the frontend.

stumpylog commented:
I'm fine with the actual feature changes and testing, but I'd prefer to get the dependency changes sorted out before we merge, so we don't need to fix `dev` hopefully.

Either a rebase off of, or merge into from, `dev`, followed by a re-locking of the dependencies seems best.  That would also confirm the docker image for this branch builds for all arches.

smseidl commented:
> I'm fine with the actual feature changes and testing, but I'd prefer to get the dependency changes sorted out before we merge, so we don't need to fix `dev` hopefully.
> 
> Either a rebase off of, or merge into from, `dev`, followed by a re-locking of the dependencies seems best. That would also confirm the docker image for this branch builds for all arches.
> 
> Edit: Also `libzbar0 poppler-utils` need to be in the image, either the base image or the main image, if they're needed for the new Python libraries.

I did try rebase from dev before changing the requirements.txt file, but that alone still failed.

gador commented:
> I'm fine with the actual feature changes and testing, but I'd prefer to get the dependency changes sorted out before we merge, so we don't need to fix `dev` hopefully.
> 
> Either a rebase off of, or merge into from, `dev`, followed by a re-locking of the dependencies seems best. That would also confirm the docker image for this branch builds for all arches.
> 

I rebased off current `dev`. I don't have the time right now to test it, and I'm still trying to get multi-arch building going. I would appreciate any testing.

> Edit: Also `libzbar0 poppler-utils` need to be in the image, either the base image or the main image, if they're needed for the new Python libraries.

Definitely. Otherwise, it will fail.

smseidl commented:
> > I'm fine with the actual feature changes and testing, but I'd prefer to get the dependency changes sorted out before we merge, so we don't need to fix `dev` hopefully.
> > Either a rebase off of, or merge into from, `dev`, followed by a re-locking of the dependencies seems best. That would also confirm the docker image for this branch builds for all arches.
> 
> I rebased off current `dev`. I don't have the time right now to test it, and I'm still trying to get multi-arch building going. I would appreciate any testing.
> 
> > Edit: Also `libzbar0 poppler-utils` need to be in the image, either the base image or the main image, if they're needed for the new Python libraries.
> 
> Definitely. Otherwise, it will fail.

@gador  - I was going to try to build directly off of your branch today, but I see the requirements.txt still has the 5.0.1 version of pikepdf.  Is there a reason that wasn't updated to 5.1.0?

gador commented:
> @gador - I was going to try to build directly off of your branch today, but I see the requirements.txt still has the 5.0.1 version of pikepdf. Is there a reason that wasn't updated to 5.1.0?

No, sorry. This must have gone lost in one of the rebase's. Fixed now.

stumpylog commented:
The base image please, as [runtime packages](https://github.com/paperless-ngx/builder/blob/main/Dockerfile.base#L24).  I'm still thinking up how to make this easier in the future.

smseidl commented:
> > @gador - I was going to try to build directly off of your branch today, but I see the requirements.txt still has the 5.0.1 version of pikepdf. Is there a reason that wasn't updated to 5.1.0?
> 
> No, sorry. This must have gone lost in one of the rebase's. Fixed now.

Even with this update, I'm still getting stuck here:
The command '/bin/sh -c ./node_modules/.bin/ng build --configuration production' returned a non-zero code: 137

smseidl commented:
> What is the actual error output above that?(presumably there is)

from the docker-compose command there is little output.  When I re-start the container by commit id, I get a bit more... but the main is about a closed pipe. 

```
# ./node_modules/.bin/ng build --configuration production
Locale data for 'be-BY' cannot be found.  Using locale data for 'be'.
WARNING [src/locale/messages.be_BY.xlf]: File target locale ('be') does not match configured locale ('be-BY')
Locale data for 'cs-CZ' cannot be found.  Using locale data for 'cs'.
WARNING [src/locale/messages.cs_CZ.xlf]: File target locale ('cs') does not match configured locale ('cs-CZ')
Locale data for 'da-DK' cannot be found.  Using locale data for 'da'.
WARNING [src/locale/messages.da_DK.xlf]: File target locale ('da') does not match configured locale ('da-DK')
Locale data for 'de-DE' cannot be found.  Using locale data for 'de'.
WARNING [src/locale/messages.de_DE.xlf]: File target locale ('de') does not match configured locale ('de-DE')
Locale data for 'es-ES' cannot be found.  Using locale data for 'es'.
Locale data for 'fr-FR' cannot be found.  Using locale data for 'fr'.
WARNING [src/locale/messages.fr_FR.xlf]: File target locale ('fr') does not match configured locale ('fr-FR')
Locale data for 'it-IT' cannot be found.  Using locale data for 'it'.
WARNING [src/locale/messages.it_IT.xlf]: File target locale ('it') does not match configured locale ('it-IT')
Locale data for 'lb-LU' cannot be found.  Using locale data for 'lb'.
WARNING [src/locale/messages.lb_LU.xlf]: File target locale ('lb') does not match configured locale ('lb-LU')
Locale data for 'nl-NL' cannot be found.  Using locale data for 'nl'.
WARNING [src/locale/messages.nl_NL.xlf]: File target locale ('nl') does not match configured locale ('nl-NL')
Locale data for 'pl-PL' cannot be found.  Using locale data for 'pl'.
WARNING [src/locale/messages.pl_PL.xlf]: File target locale ('pl') does not match configured locale ('pl-PL')
Locale data for 'pt-BR' cannot be found.  Using locale data for 'pt'.
Locale data for 'ro-RO' cannot be found.  Using locale data for 'ro'.
WARNING [src/locale/messages.ro_RO.xlf]: File target locale ('ro') does not match configured locale ('ro-RO')
Locale data for 'ru-RU' cannot be found.  Using locale data for 'ru'.
⠋ Generating browser application bundles (phase: setup)...WARNING [src/locale/messages.ru_RU.xlf]: File target locale ('ru') does not match configured locale ('ru-RU')
Locale data for 'sl-SI' cannot be found.  Using locale data for 'sl'.
WARNING [src/locale/messages.sl_SI.xlf]: File target locale ('sl') does not match configured locale ('sl-SI')
Locale data for 'sr-CS' cannot be found.  Using locale data for 'sr'.
Locale data for 'sv-SE' cannot be found.  Using locale data for 'sv'.
Locale data for 'tr-TR' cannot be found.  Using locale data for 'tr'.
WARNING [src/locale/messages.tr_TR.xlf]: File target locale ('tr') does not match configured locale ('tr-TR')
Locale data for 'zh-CN' cannot be found.  Using locale data for 'zh'.
⠏ Generating browser application bundles (phase: building)...Killed
# fatal error: too many writes on closed pipe

goroutine 6 [running]:
runtime.throw({0x894f2, 0x1e})
        /usr/local/go/src/runtime/panic.go:1198 +0x7 fp=0x434ee0 sp=0x434eb8 pc=0x12000007
os.sigpipe()
        /usr/local/go/src/runtime/os_js.go:143 +0x2 fp=0x434ef8 sp=0x434ee0 pc=0x137c0002
os.epipecheck(...)
        /usr/local/go/src/os/file_unix.go:196
os.(*File).Write(0x40c020, {0x616580, 0x21, 0x40})
        /usr/local/go/src/os/file.go:184 +0x25 fp=0x434f80 sp=0x434ef8 pc=0x15b90025
main.runService.func1(0x40f9e0)
        /Users/evan/dev/esbuild/cmd/esbuild/service.go:114 +0xa fp=0x434fd8 sp=0x434f80 pc=0x1dbc000a
runtime.goexit()
        /usr/local/go/src/runtime/asm_wasm.s:431 +0x1 fp=0x434fe0 sp=0x434fd8 pc=0x13d10001
created by main.runService
        /Users/evan/dev/esbuild/cmd/esbuild/service.go:108 +0xe

goroutine 1 [chan receive]:
syscall.fsCall({0x7c3f2, 0x4}, {0x45fb98, 0x5, 0x5})
        /usr/local/go/src/syscall/fs_js.go:521 +0x13
syscall.Read(0x0, {0x4ca000, 0x4000, 0x4000})
        /usr/local/go/src/syscall/fs_js.go:389 +0xc
internal/poll.ignoringEINTRIO(...)
        /usr/local/go/src/internal/poll/fd_unix.go:582
internal/poll.(*FD).Read(0x42a060, {0x4ca000, 0x4000, 0x4000})
        /usr/local/go/src/internal/poll/fd_unix.go:163 +0x4c
os.(*File).read(...)
        /usr/local/go/src/os/file_posix.go:32
os.(*File).Read(0x40c018, {0x4ca000, 0x4000, 0x4000})
        /usr/local/go/src/os/file.go:119 +0x10
main.runService(0x1)
        /Users/evan/dev/esbuild/cmd/esbuild/service.go:141 +0x23
main.main()
        /Users/evan/dev/esbuild/cmd/esbuild/main.go:207 +0x8

goroutine 7 [waiting]:
runtime.gopark(0x0, 0x0, 0x0, 0x0, 0x1)
        /usr/local/go/src/runtime/proc.go:366 +0x27
runtime.handleEvent()
        /usr/local/go/src/runtime/lock_js.go:250 +0x1b
runtime.goexit()
        /usr/local/go/src/runtime/asm_wasm.s:431 +0x1

goroutine 8 [chan receive]:
main.(*serviceType).sendRequest(0x40f9e0, {0x3dde0, 0x4fda70})
        /Users/evan/dev/esbuild/cmd/esbuild/service.go:203 +0xe
main.runService.func2(0x40f9e0)
        /Users/evan/dev/esbuild/cmd/esbuild/service.go:132 +0x3
created by main.runService
        /Users/evan/dev/esbuild/cmd/esbuild/service.go:129 +0x1e

```

shamoon commented:
Definitely a little esoteric but I’m not convinced the issue is about an issue with our particular depencies. Googling a bit found: https://github.com/evanw/esbuild/issues/1662#issuecomment-1001763776 (maybe you saw that already)

smseidl commented:
> Definitely a little esoteric but I’m not convinced the issue is about an issue with our particular depencies. Googling a bit found: [evanw/esbuild#1662 (comment)](https://github.com/evanw/esbuild/issues/1662#issuecomment-1001763776) (maybe you saw that already)

I ran across that at one point, but at the time felt it was my user error than anything.  I will try that change in the near future.   The one thing that doesn't explain is what it appears to fix when I remove that `--no-optional` option.

gador commented:
rebased (again) to fix merge conflict with dev

smseidl commented:
> > Definitely a little esoteric but I’m not convinced the issue is about an issue with our particular depencies. Googling a bit found: [evanw/esbuild#1662 (comment)](https://github.com/evanw/esbuild/issues/1662#issuecomment-1001763776) (maybe you saw that already)
> 
> I ran across that at one point, but at the time felt it was my user error than anything. I will try that change in the near future. The one thing that doesn't explain is what it appears to fix when I remove that `--no-optional` option.

This appeared to work.  I'm going to do another clean build later tonight to confirm it.

smseidl commented:
> > > Definitely a little esoteric but I’m not convinced the issue is about an issue with our particular depencies. Googling a bit found: [evanw/esbuild#1662 (comment)](https://github.com/evanw/esbuild/issues/1662#issuecomment-1001763776) (maybe you saw that already)
> > 
> > 
> > I ran across that at one point, but at the time felt it was my user error than anything. I will try that change in the near future. The one thing that doesn't explain is what it appears to fix when I remove that `--no-optional` option.
> 
> This appeared to work. I'm going to do another clean build later tonight to confirm it.

Here is what I was able to get built.  I haven't tested anywhere other than the build host currently (and only slightly).  I will be trying more testing today:
https://hub.docker.com/r/baldgolfer53/paperless-ngx_featurebuilds

gador committed:
add split logic to consume_file

Signed-off-by: florian on nixos (Florian Brandes) <florian.brandes@posteo.de>

smseidl commented:
>
> Here is what I was able to get built. I haven't tested anywhere other than the build host currently (and only slightly). I will be trying more testing today: https://hub.docker.com/r/baldgolfer53/paperless-ngx_featurebuilds

This image is not working in my testing tonight.  It's missing the zbar shared library.  This did not prevent the image from getting built, but isn't allowing for any processing of PDF docs.

```
20:43:17 [Q] INFO Enqueued 1
20:43:17 [Q] INFO Process-1:10 processing [Into th Breach Poster.pdf]
Process Process-1:10:
Traceback (most recent call last):
  File "/usr/local/lib/python3.9/pydoc.py", line 439, in safeimport
    module = __import__(path)
  File "/usr/src/paperless/src/documents/tasks.py", line 25, in <module>
    from pyzbar import pyzbar
  File "/usr/local/lib/python3.9/site-packages/pyzbar/pyzbar.py", line 7, in <module>
    from .wrapper import (
  File "/usr/local/lib/python3.9/site-packages/pyzbar/wrapper.py", line 151, in <module>
    zbar_version = zbar_function(
  File "/usr/local/lib/python3.9/site-packages/pyzbar/wrapper.py", line 148, in zbar_function
    return prototype((fname, load_libzbar()))
  File "/usr/local/lib/python3.9/site-packages/pyzbar/wrapper.py", line 127, in load_libzbar
    libzbar, dependencies = zbar_library.load()
  File "/usr/local/lib/python3.9/site-packages/pyzbar/zbar_library.py", line 65, in load
    raise ImportError('Unable to find zbar shared library')
ImportError: Unable to find zbar shared library

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/local/lib/python3.9/multiprocessing/process.py", line 315, in _bootstrap
    self.run()
  File "/usr/local/lib/python3.9/multiprocessing/process.py", line 108, in run
    self._target(*self._args, **self._kwargs)
  File "/usr/local/lib/python3.9/site-packages/django_q/cluster.py", line 424, in worker
    f = pydoc.locate(f)
  File "/usr/local/lib/python3.9/pydoc.py", line 1718, in locate
    nextmodule = safeimport('.'.join(parts[:n+1]), forceload)
  File "/usr/local/lib/python3.9/pydoc.py", line 454, in safeimport
    raise ErrorDuringImport(path, sys.exc_info())
pydoc.ErrorDuringImport: problem in documents.tasks - ImportError: Unable to find zbar shared library
```

gador commented:
Yep. Thats why:

> The base image please, as [runtime packages](https://github.com/paperless-ngx/builder/blob/main/Dockerfile.base#L24). I'm still thinking up how to make this easier in the future.

I still need to add it there. Before, it was part of the Dockerfile, before the build process changed

gador commented:
> The base image please, as [runtime packages](https://github.com/paperless-ngx/builder/blob/main/Dockerfile.base#L24). I'm still thinking up how to make this easier in the future.

@stumpylog I added the requirements and the base image rebuilt successfully. 
I also did another rebase, due to changes in requirements.txt. Does this new CI build of the feature branch use the newly build image from the builder?

gador commented:
ok. I added the requirements to the base build, added a new version to the base build and did a rebuild here.

The image can be found at `ghcr.io/paperless-ngx/paperless-ngx:feature-barcode-splitter`

smseidl reviewed (commented):
I am not a python expert so not doing this as an inline comment cause there probably is valid reason... but I can't understand the need for the `else` statement in separate_pages method.  It looks like this is just getting the first PDF from original file.  Wouldn't the first iteration of the following for loop accomplish the same thing?

smseidl commented:
> ok. I added the requirements to the base build, added a new version to the base build and did a rebuild here.
> 
> The image can be found at `ghcr.io/paperless-ngx/paperless-ngx:feature-barcode-splitter`

I was able to get this running and processed my first document with the PATCHT file :-).  I will be running more real volume thru the next couple days.

gador committed:
fix conditional in separate_pages
add additional test for separate_pages

Signed-off-by: Florian Brandes <florian.brandes@posteo.de>

gador commented:
> I am not a python expert so not doing this as an inline comment cause there probably is valid reason... but I can't understand the need for the `else` statement in separate_pages method. It looks like this is just getting the first PDF from original file. Wouldn't the first iteration of the following for loop accomplish the same thing?

Well, the `else` statement is just part of the check, whether there're pages to even split upon.
For the second part of the question, why:

I found it easier that way. 

The first part after the  `else` statement just iterates from zero to the first element of the list `pages_to_split_on`
If I don't do this first and try to do this in the loop after, I get a problem here: 
```
for page in range(page_number + 1, next_page):
                logger.debug(f"page_number: {str(page_number)} next_page: {str(next_page)}")
                dst.pages.append(pdf.pages[page])
```

I'd have to create an additional conditional check, whether the `count` is zero and set the `page_number`  to zero (since we would iterate from the beginning of the document)

I tried it at first to have less redundency in the code, but found it more complicated to understand later on. I tried to write easy and self-explaining code, so I can understand it later on ;-)

https://github.com/paperless-ngx/paperless-ngx/blob/cdf5602dfbea88c548cc1e599738e260d98cfa15/src/documents/tasks.py#L120-L156

I did however found a bug through your question. The for loop in L140-156 should be indented one level to be on the `else` statement you mentioned. Otherwise i'll try to iterate, even if there are no pages to split.

smseidl commented:
@gador Thanks for the explanation that was about what I was expecting.  I will be pulling and testing this over the weekend.  Thanks for all the hard work on this.

smseidl commented:
@gador  - one very minor thing I'm seeing in my testing tonight is some "duplicate logging" (not really duplicate, just repeative in the for loop).  Certainly not a stopper, but something that maybe should be changed to 'debug' level?

```
[2022-04-09 23:04:42,082] [DEBUG] [paperless.tasks] Count: 7 page_number: 47
[2022-04-09 23:04:42,083] [DEBUG] [paperless.tasks] page_number: 47 next_page: 68
[2022-04-09 23:04:42,087] [DEBUG] [paperless.tasks] page_number: 47 next_page: 68
[2022-04-09 23:04:42,114] [DEBUG] [paperless.tasks] page_number: 47 next_page: 68
[2022-04-09 23:04:42,131] [DEBUG] [paperless.tasks] page_number: 47 next_page: 68
[2022-04-09 23:04:42,134] [DEBUG] [paperless.tasks] page_number: 47 next_page: 68
[2022-04-09 23:04:42,145] [DEBUG] [paperless.tasks] page_number: 47 next_page: 68
[2022-04-09 23:04:42,152] [DEBUG] [paperless.tasks] page_number: 47 next_page: 68
[2022-04-09 23:04:42,172] [DEBUG] [paperless.tasks] page_number: 47 next_page: 68
[2022-04-09 23:04:42,320] [DEBUG] [paperless.tasks] page_number: 47 next_page: 68
```

gador commented:
It seems you have debugging already turned on 😉


For the error in zlib:

I found https://github.com/mchehab/zbar/issues/177 which seems to be the same error and I found an unanswered stackOverflow answer.
Seems like noone has an answer to that.
It could be related to non-x86_64-linux systems..

Does it hinder the Barcode detection?

smseidl commented:
@gador  one other item I noticed - the progress bar on the dashboard got stuck for the document I uploaded containing the separator pages.

![image](https://user-images.githubusercontent.com/23556348/162601866-25b192e7-2e60-4be4-9d38-fa313c362423.png)

smseidl commented:
> It seems you have debugging already turned on 😉
> 
> For the error in zlib:
> 
> I found [mchehab/zbar#177](https://github.com/mchehab/zbar/issues/177) which seems to be the same error and I found an unanswered stackOverflow answer. Seems like noone has an answer to that. It could be related to non-x86_64-linux systems..
> 
> Does it hinder the Barcode detection?

How did I managed to turn that on? :-o  I will have to look at that.  I look at to many similarly formatted logs at work and don't even look at the level.

No, that log text did not seem to hinder processing at all.

gador commented:
> @gador  one other item I noticed - the progress bar on the dashboard got stuck for the document I uploaded containing the separator pages.
> 
> ![image](https://user-images.githubusercontent.com/23556348/162601866-25b192e7-2e60-4be4-9d38-fa313c362423.png)
> 

That is weird. It should finish once all documents are processed.
Are you sure that all documents contained in the scan completed? What do the logs say? Do the progress bars of the sub-documents finish correctly?

smseidl commented:
> > @gador  one other item I noticed - the progress bar on the dashboard got stuck for the document I uploaded containing the separator pages.
> > ![image](https://user-images.githubusercontent.com/23556348/162601866-25b192e7-2e60-4be4-9d38-fa313c362423.png)
> 
> That is weird. It should finish once all documents are processed. Are you sure that all documents contained in the scan completed? What do the logs say? Do the progress bars of the sub-documents finish correctly?

Yes, there were 13 documents in it, and I had files 0_* thru 12_* in Paperless when done.  The sub-document progress bars came and went as well.  The progress bar for the original doc did go away with a F5 Refresh.

gador commented:
@smseidl thanks for your tests!
If no show-stopper bugs appear, I'll go ahead and merge tomorrow into dev

fabicodes commented:
First of all, thanks for the feature!
One idea I have, that could be based on it, is this. One could create support for barcodes with sequential number. The number could be treated as an ASN and when the barcode is encountered, the document could be split at that point.
The barcodes could be created with a label printer and then sticked on the documents to be scanned.

gador commented:
yes, there is already a discussion about it [here](https://github.com/paperless-ngx/paperless-ngx/discussions/648)
